backends/cuda/ceed-cuda-qfunctioncontext.c

```diff
@@ -57,198 +74,278 @@ static inline int CeedQFunctionContextSyncD2H_Cuda(
   CeedQFunctionContext_Cuda *impl;
   ierr = CeedQFunctionContextGetBackendData(ctx, &impl); CeedChkBackend(ierr);
 
+  if (!impl->d_data)
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_BACKEND,
+                     "No valid device data to sync to host");
+  // LCOV_EXCL_STOP
+
+  if (impl->h_data_borrowed) {
+    impl->h_data = impl->h_data_borrowed;
+  } else if (impl->h_data_owned) {
+    impl->h_data = impl->h_data_owned;
+  } else {
+    ierr = CeedMalloc(bytes(ctx), &impl->h_data_owned);
+    CeedChkBackend(ierr);
+    impl->h_data = impl->h_data_owned;
+  }
+
   ierr = cudaMemcpy(impl->h_data, impl->d_data, bytes(ctx),
                     cudaMemcpyDeviceToHost); CeedChk_Cu(ceed, ierr);
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Sync data of type
+//------------------------------------------------------------------------------
+static inline int CeedQFunctionContextSync_Cuda(
+  const CeedQFunctionContext ctx, CeedMemType mtype) {
+  switch (mtype) {
+  case CEED_MEM_HOST: return CeedQFunctionContextSyncD2H_Cuda(ctx);
+  case CEED_MEM_DEVICE: return CeedQFunctionContextSyncH2D_Cuda(ctx);
+  }
+  return CEED_ERROR_UNSUPPORTED;
+}
+
+//------------------------------------------------------------------------------
+// Set all pointers as invalid
+//------------------------------------------------------------------------------
+static inline int CeedQFunctionContextSetAllInvalid_Cuda(
+  const CeedQFunctionContext ctx) {
+  int ierr;
+  CeedQFunctionContext_Cuda *impl;
+  ierr = CeedQFunctionContextGetBackendData(ctx, &impl); CeedChkBackend(ierr);
+
+  impl->h_data = NULL;
+  impl->d_data = NULL;
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Check if ctx has valid data
+//------------------------------------------------------------------------------
+static inline int CeedQFunctionContextHasValidData_Cuda(
+  const CeedQFunctionContext ctx, bool *has_valid_data) {
+  int ierr;
+  CeedQFunctionContext_Cuda *impl;
+  ierr = CeedQFunctionContextGetBackendData(ctx, &impl); CeedChkBackend(ierr);
+
+  *has_valid_data = !!impl->h_data || !!impl->d_data;
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Check if data of given type needs sync
+//------------------------------------------------------------------------------
+static inline int CeedQFunctionContextNeedSync_Cuda(
+  const CeedQFunctionContext ctx, CeedMemType mtype, bool *need_sync) {
```

Comment:
I really like this way of checking whether we need to sync (and in the vector stuff, too).  I think it's easier to follow than what we had before.